[DCP] Introduce process based async checkpointing

meetv18 opened the pull request:
Summary:
### Context
Background checkpoint upload thread interfering with trainer thread:

In [async save API](https://github.com/pytorch/pytorch/blob/main/torch/distributed/checkpoint/state_dict_saver.py#L239-L248), the background thread spends a considerable amount of time on CPU-bound tasks (pickling/unpickling several metada objects a.k.a SavePlans) on rank0 during the collective operation; this kind of asymmetric computation heavily contends for GIL with the trainer thread causing GPU util to suffer significantly for the E2E checkpoint duration. 

### Solution:
Introduce async save via a checkpoint daemon process. This daemon process will be created once (during the first save attempt) and can serve async checkpoint requests for the remainder of training lifetime.

Test Plan: Added E2E UTs for process based async save. 

Differential Revision: D69272583




cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @LucasLLC @mhorowitz @pradeepfn @ekr0

saumishr commented on torch/distributed/checkpoint/_save_process.py:
```diff
@@ -0,0 +1,288 @@
+            map_fun=get_free_port,
+        )
+
+class _AsyncCheckpointProcess:
```
Not a blocker for the first iteration and quick experimentations we wanna do, but lets make this interface a bit more generic as a general process based executor. We can leverage the same for the staging as well to enable zero overhead staging, potentially.

saumishr commented on torch/distributed/checkpoint/state_dict_saver.py:
```diff
@@ -255,6 +299,41 @@ def async_save(
+    if isinstance(storage_writer, AsyncProcessStager):
+        staged_state_dict, local_sync_state = storage_writer.stage(state_dict)
+    else:  # provides bwc for storage_writers not implementing AsyncStager
+        staged_state_dict = _create_cpu_state_dict(state_dict, share_memory=True)
```
We can probably keep the code DRY between _async_save_process_based and _async_save_thread_based as long as we do a dependency injection of an executor. Underlying implementation can be Thread based or process based depending on a config.

saumishr commented on torch/distributed/checkpoint/state_dict_saver.py:
```diff
@@ -4,26 +4,40 @@
+__all__ = ["save_state_dict", "save", "async_save", "AsyncCheckpointerType"]
+
+class AsyncCheckpointerType(Enum):
+    """Enum for async checkpoint type."""
```
`nit`: Enum for async `checkpointer` type.

saumishr commented on torch/distributed/checkpoint/state_dict_saver.py:
```diff
@@ -255,6 +299,41 @@ def async_save(
+    """
+
+    local_sync_state: Optional[LocalSyncState] = None
+    if isinstance(storage_writer, AsyncProcessStager):
```
AsyncProcessStager doesn't seem very intuitive to understand given that currently process is being used for upload and not staging. 

We probably only need the following components:
- Stager
- CheckpointExecutor [Thread based, Process based]

Depending on whether the user has set the synchronize_after_execute flag, Stager becomes sync or async. 

For async staging, it should be fine to use process based CheckpointExecutor by default. Given that threads may affect the QPS in the forward step, I don't think there will be any preference given to a thread based CheckpointExecutor.

We introduce a config for upload_executor: [thread, process]. Its an internal detail and probably we can flip the default to process based and keep threads as optional in case a user wants to continue using threads.

gkroiz commented on torch/distributed/checkpoint/state_dict_saver.py:
```diff
@@ -256,6 +302,42 @@ def async_save(
+
+    local_sync_state: Optional[LocalSyncState] = None
+    if isinstance(storage_writer, AsyncProcessStager):
+        staged_state_dict, local_sync_state = storage_writer.stage(state_dict)
```
I was messing around with this code in a local environment and run into an error with this line. The default `storage_writer.stage(state_dict)` for `async_save` returns 1 object, not 2

gkroiz replied:
I no longer run into this issue with the latest changes.

saumishr commented on torch/distributed/checkpoint/_async_executor.py:
```diff
@@ -0,0 +1,379 @@
+            raise
+
+        try:
+            while True:
```
Is this an acceptable pattern for the process to wait on a message? This loop will be continuously consuming the CPU. Before making the process based implementation a default, it would be good to understand the cost here.

saumishr commented on torch/distributed/checkpoint/_async_executor.py:
```diff
@@ -0,0 +1,379 @@
+        planner: Optional[SavePlanner] = None,
+        process_group: Optional[dist.ProcessGroup] = None,
+    ) -> Future:
+        """
```
We should probably separate the interface and concrete implementations in different files.

saumishr commented on torch/distributed/checkpoint/_async_executor.py:
```diff
@@ -0,0 +1,379 @@
+        """
+        Execute the checkpoint save request asynchronously.
+
+        This method is intended to be used as an abastraction for
```
`nit` abastraction -> abstraction

saumishr commented on torch/distributed/checkpoint/_async_executor.py:
```diff
@@ -0,0 +1,379 @@
+        """
+
+
+class _ProcessBasedAsyncCheckpointExecutor(_AsyncCheckpointExecutor):
```
Same here. Move out thread based and process based executors in their own files. It would make the code more readable for the OSS users and also debug as needed.

saumishr commented on torch/distributed/checkpoint/staging.py:
```diff
@@ -47,7 +53,9 @@ def should_synchronize_after_execute(self) -> bool:
+        self, state_dict: STATE_DICT_TYPE
+    ) -> Union[STATE_DICT_TYPE, Tuple[STATE_DICT_TYPE, StagingSyncState]]:
         """
         Returns a "staged" copy of `state_dict`. The expectation of the staged copy is that it is
```
Update the documentation on the StagingSyncState here.

saumishr commented on torch/distributed/checkpoint/staging.py:
```diff
@@ -56,7 +64,9 @@ def stage(self, state_dict: STATE_DICT_TYPE) -> STATE_DICT_TYPE:
-    def synchronize_staging(self) -> None:
+    def synchronize_staging(
+        self, staging_sync_state: Optional[StagingSyncState] = None
+    ) -> None:
```
Same here. Lets update the function doc as well

saumishr commented on torch/distributed/checkpoint/state_dict_saver.py:
```diff
@@ -230,29 +244,48 @@ def async_save(
+        ):
+            warnings.warn(
+                "The function definition for StorageWriter.synchronize_staging has been updated"
+                " to include the staging_sync_state argument. Please update your implementation"
```
It may be a bit less intuitive for a user to understand what staging_sync_state is about. Could we please add some documentation on it.

saumishr commented on torch/distributed/checkpoint/state_dict_saver.py:
```diff
@@ -230,29 +244,48 @@ def async_save(
+            )
+            storage_writer.synchronize_staging()  # type: ignore[call-arg]
+        else:
+            storage_writer.synchronize_staging(staging_sync_state)
```
Lets discuss on the zero overhead checkpointing strategy here and figure out what would be the return type for the user when `should_synchronize_after_execute` is false. Specifically, if we will  return a cuda stream, a future?

saumishr reviewed (commented):
Looks good overall! A few minor comments on the code structure. We can get this landed after that.

gkroiz commented on torch/distributed/checkpoint/state_dict_saver.py:
```diff
@@ -230,29 +244,48 @@ def async_save(
-    f: Future = executor.submit(
-        save,
-        staged_state_dict,
+    executor: _AsyncCheckpointExecutor = (
```
When I'm experimenting with this code, I notice that the first time saving async checkpoint via processes that it takes some time to spawn the processes, even with small model configurations (for example, model size 36M on 8H100s will create subprocesses 10-15 seconds after calling async_save. I'm curious do you also see this behavior? If so, could we optimize here, perhaps spawn subprocesses while defining staged_state_dict?

gkroiz replied:
Adding details from online sync, most of the time spent during the initialization of the checkpoint subprocesses is when creating the subprocess' distributed group. What might make sense here is to define an API that allows you to create the checkpoint subprocesses while training also initializes.

daulet-askarov replied:
+1. I think we can spin off a Future initing the async sub-process and the sub-process starting to establish upload-side ProcessGroup between ranks early, maybe in the BaseStorageWriter init itself. By the time of the first save everything should be ready, and we can just call future.result() in the beginning of async save for safety.

daulet-askarov commented on torch/distributed/checkpoint/_async_executor.py:
```diff
@@ -0,0 +1,379 @@
+                master_addr = _get_fq_hostname()
+            return master_addr
+
+        self.tcp_store_master_addr = dist_wrapper.broadcast(
+            step="get_master_addr",
+            map_fun=get_master_addr,
+        )
+        self.tcp_store_master_port = dist_wrapper.broadcast(
+            step="get_free_port",
+            map_fun=get_free_port,
+        )
```
This seems wasteful to do 2 broadcasts in a row. Why not define compose get_master_addr and get_free_port into a single function and broadcast its output pair tuple once from rank 0 to all the other ranks?

daulet-askarov commented on torch/distributed/checkpoint/_async_executor.py:
```diff
@@ -0,0 +1,379 @@
+                global _CHECKPOINT_PROCESS
+                _CHECKPOINT_PROCESS = _AsyncCheckpointProcess(pg_init_info=pg_init_info)
+
+            create_checkpoint_daemon_process()
```
What not put create_checkpoint_daemon_process() call site inside BaseStorageWriter `__init__` and make it asynchronous by creating a async_proccess_creation_future? There will be plenty of time between BaseStorageWriter.__init__ and the first save, so the first save then doesn't need to suffer the additional latency, you can just call async_proccess_creation_future.result() in the beginning of _execute_save_impl to enforce that the process got inited.

daulet-askarov commented on torch/distributed/checkpoint/_async_executor.py:
```diff
@@ -0,0 +1,379 @@
+        if _CHECKPOINT_PROCESS is None:
+            # Find a free port on coordinator rank and broadcast
+            # to all ranks.
+            pg_init_info = _ProcessGroupInitInfo(process_group)
```
This call's latency can also be hidden into a separate thread, the same one which we can introduce for initing AsyncCheckpointProcess.

saumishr reviewed (approved):
Looks good! Please fix the Lint errors before landing.

meetv18 committed:
[DCP] Introduce process based async checkpointing (#147039)

Summary:

### Context
Background checkpoint upload thread interfering with trainer thread:

In [async save API](), the background thread spends a considerable amount of time on CPU-bound tasks (pickling/unpickling several metada objects a.k.a SavePlans) on rank0 during the collective operation; this kind of asymmetric computation heavily contends for GIL with the trainer thread causing GPU util to suffer significantly for the E2E checkpoint duration. 

### Solution:
Introduce async save via a checkpoint daemon process. This daemon process will be created once (during the first save attempt) and can serve async checkpoint requests for the remainder of training lifetime.

Test Plan: E2E UTs

Reviewed By: saumishr

Differential Revision: D69272583

teja-rao commented:
A bit late to the party but why do we need thread based checkpointer given process based one out performs thread based impl?

Also, switch to using multiprocessing.Pipe from queue. so the recv on the pipe does not block if the sub-process died abruptly. subprocess can die abruptly from a SIGTERM/SIGINT/SIGKILL etc.
